### PR TITLE
[#1050] Add context menu actions to ReferenceSearchViewPage

### PR DESCRIPTION
**Pull request by kthoms**
- add Copy action
- add Remove Selected Matches action
- bind actions to key events

The two new commands can be selected from the context menu in the search result view or directly activated via keyboard with `CMD+C` or `DELETE`.

### CODE REVIEW

**Comment by kthoms** [comment]
<div class="cloudapp-embed" data-slug="d4e567f5e9d8"><a href="https://cl.ly/d4e567f5e9d8">Screen Recording 2019-04-16 at 11.43 PM.mov</a><script async src="https://embed.cl.ly/embedded.gz.js" charset="utf-8"></script></div>

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java [hunk 1]
```diff
@@ -162,6 +173,23 @@ public void createControl(Composite parent) {
...
+		menu.addMenuListener(mgr -> {
+			fillContextMenu(mgr);
+			part.fillContextMenu(mgr);
+		});
+		
+		viewer.getControl().setMenu(menu.createContextMenu(viewer.getControl()));
+		viewer.getControl().addKeyListener(KeyListener.keyPressedAdapter(e -> {
+			if (e.keyCode == SWT.DEL) {
+				removeSelectedMatchesAction.run();
+			} else if ((e.stateMask|SWT.COMMAND)!=0 && e.keyCode=='c') {
```
Comment (cdietrich) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: formatting

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java [hunk 2]
```diff
@@ -253,4 +286,12 @@ public TreeViewer getViewer() {
 		return viewer;
 	}
 
+	ReferenceSearchResultContentProvider getContentProvider() {
```
Comment (cdietrich) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: why not protected?

**Comment by cdietrich** [comment]
wonder if we have tests for this that could be extended

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java (hunk 1 above)
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: ok

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java (hunk 2 above)
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: because package private visibility is enough

Comment (cdietrich) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: but is will prevent any customization

Comment (cdietrich) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: @szarnekow what do you think?

**Comment by kthoms** [comment]
> wonder if we have tests for this that could be extended

Did not find any. Would it be worth to add some? And if so, should this be done with this ticket? Should they be added with 2.18 or is later enough?

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java [hunk 3]
```diff
@@ -255,4 +255,18 @@ public boolean apply(IReferenceDescription input) {
...
+	
+	public void remove (ReferenceSearchViewTreeNode... nodes) {
+		for (ReferenceSearchViewTreeNode node : nodes) {
+			if (!rootNodes.remove(node)) { // if not top-level node then recurse into children
+				rootNodes.forEach(n -> remove(node, n));
+			}
+		}
+	}
+	
+	private void remove (ReferenceSearchViewTreeNode toRemove, ReferenceSearchViewTreeNode from) {
```
Comment (szarnekow) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java]: Should be protected

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/messages.properties [hunk 4]
```diff
@@ -14,3 +14,7 @@ ReferenceSearchViewPageActions_showNextMatch=Show Next Match
 ReferenceSearchViewPageActions_showNextMatch_tooltip=Show Next Match
 ReferenceSearchViewPageActions_showPreviousMatch=Show Previous Match
 ReferenceSearchViewPageActions_showPreviousMatch_tooltip=Show Previous Match
+ReferenceSearchViewPageActions_copy=Copy
+ReferenceSearchViewPageActions_copy_tooltip=Copy
+ReferenceSearchViewPageActions_removeSelectedMatches=Remove Selected Matches
```
Comment (cdietrich) [inline org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/messages.properties]: what about remove all matches as well?

**Comment by cdietrich** [comment]
Needs to be rebased. also address review findings

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java (hunk 3 above)
Comment (szarnekow) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java]: Ping

Comment (szarnekow) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java]: Pong @kthoms

**Comment by cdietrich** [comment]
ping @kthoms

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java (hunk 3 above)
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java]: OK, will change

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java [hunk 5]
```diff
@@ -255,4 +255,18 @@ public boolean apply(IReferenceDescription input) {
...
+		from.removeChild(toRemove);
+		from.getChildren().forEach(childNode -> remove(toRemove, childNode));
```
[the remark this answers is not in the document]
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java]: Some hundred maybe? Do you fear locking the UI here? Should I put this into a background job? I suppose it does not need to be performed on the UI thread.
This is performed when you select nodes in the search result tree and remove them.

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java [hunk 6]
```diff
@@ -233,6 +261,11 @@ protected void fillToolbar(IToolBarManager tbm) {
 		tbm.appendToGroup(IContextMenuConstants.GROUP_VIEWER_SETUP, collapseAllAction);
 	}
 
+	private void fillContextMenu(IMenuManager mgr) {
```
[the remark this answers is not in the document]
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: OK

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java (hunk 2 above)
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPage.java]: OK, will change

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPageActions.java [hunk 7]
```diff
@@ -80,4 +93,56 @@ public void run() {
...
+			this.page = page;
+		}
+
+		@Override
+		public void run() {
+			TreeViewer viewer = page.getViewer();
+			Clipboard clipboard = new Clipboard(viewer.getControl().getDisplay());
+			
+			ITreeSelection selection = page.getViewer().getStructuredSelection();
+			@SuppressWarnings("unchecked")
```
[the remark this answers is not in the document]
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPageActions.java]: `org.eclipse.jface.viewers.IStructuredSelection#toList` is not generic

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPageActions.java [hunk 8]
```diff
@@ -80,4 +93,56 @@ public void run() {
...
+			Object data = selection.toList().stream()
+				.map(sel -> page.getLabelProvider().getText(sel))
+				.collect(joining(System.lineSeparator()));
+			clipboard.setContents(new Object[] {data}, new Transfer[]{TextTransfer.getInstance()});
```
[the remark this answers is not in the document]
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchViewPageActions.java]: I was not aware that this is possible. Interesting idea.

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/messages.properties (hunk 4 above)
Comment (kthoms) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/messages.properties]: I'll remove them.

File: org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (szarnekow) [reply org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/findrefs/ReferenceSearchResultContentProvider.java]: No, I wasn't thinking about a background job. If it's only a few elements, the algorithm here looks ok. I'll mark this as resolved.

**Comment by cdietrich** [comment]
@szarnekow @kthoms what is the state of this pr?

**Pull request merged**